lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -200,16 +201,47 @@ def safe_add_concurrent_partitioned_index(
 
     _validate_index_name!(parent_index)
 
+    # Validate index names upfront to avoid unnecessary recursion
     child_tables_with_metadata = _partitions_for_table(parent_schema, parent_table).map do |child_schema, child_table|
-      raise PgHaMigrations::InvalidMigrationError, "Partitioned table #{parent_table.inspect} contains sub-partitions" if _partitioned_table?(child_schema, child_table)
-
       child_index = "#{child_table}_#{name_suffix}"
 
       _validate_index_name!(child_index)
 
       [child_schema, child_table, child_index]
     end
 
+    # When sub-partitioning detected:
+    #   - recursively create indexes when recurse is true
+    #   - do nothing when if_not_exists is true and index already valid
+    #   - otherwise raise helpful error
+    child_tables_with_metadata.each do |child_schema, child_table, child_index|
```

Comment:
I'm not sure I'm following this one. In both scenarios, if `recurse: true` the result will be a valid index on the parent and if `recurse: false` it will raise an error that sub-partitioning was detected